contracts/libraries/OracleLibrary.sol

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity >=0.5.0 <0.8.0;
+
+import '@uniswap/v3-core/contracts/libraries/FullMath.sol';
+import '@uniswap/v3-core/contracts/libraries/TickMath.sol';
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+import '@uniswap/v3-core/contracts/libraries/LowGasSafeMath.sol';
+import '../libraries/PoolAddress.sol';
+
+/// @title Oracle library
+/// @notice Provides functions to integrate with V3 pool oracle
+library OracleLibrary {
+    /// @notice Fetches time-weighted average tick using Uniswap V3 oracle
+    /// @param factory Address of Uniswap V3 factory contract
+    /// @param token0 Address of an ERC20 token contract
+    /// @param token1 Address of an ERC20 token contract
+    /// @param fee Fee amount that correspond to the pool that we want to observe
+    /// @param period Number of seconds in the past to start calculating time-weighted average
+    /// @return tick The time-weighted average tick from (block.timestamp - period) to block.timestamp
+    function consult(
+        address factory,
+        address token0,
+        address token1,
+        uint24 fee,
+        uint32 period
+    ) internal view returns (int24 tick) {
+        require(period != 0, 'BP');
+        IUniswapV3Pool oracle =
+            IUniswapV3Pool(PoolAddress.computeAddress(factory, PoolAddress.getPoolKey(token0, token1, fee)));
+
+        uint32[] memory secondAgos = new uint32[](2);
+        secondAgos[0] = period;
+        secondAgos[1] = 0;
+
+        (int56[] memory tickCumulatives, ) = oracle.observe(secondAgos);
+        int56 tickCumulativesDelta = tickCumulatives[1] - tickCumulatives[0];
+
+        tick = int24(tickCumulativesDelta / period);
+
+        // Always round to negative infinity
+        tick = (tick < 0 && (tickCumulativesDelta % period != 0)) ? tick - 1 : tick;
```

Comment:
could you maybe avoid an op by doing `if (tick < 0 && (tickCumulativesDelta % period != 0)) tick--;`

Reply:
needs to check the numerator < 0, rather than the quotient